src/QueryPipeline/RemoteQueryExecutor.h

```diff
@@ -94,12 +94,18 @@ class RemoteQueryExecutor
     /// Query is resent to a replica, the query itself can be modified.
     std::atomic<bool> resent_query { false };
 
+    struct ParallelReplicasToken {};
+    struct Nothing {};
+    using ReadResult = std::variant<Block, ParallelReplicasToken, int, Nothing>;
```

Comment:
I think this is ok.

But, probably, a bit better to make EnumClass + struct ...

Reply:
Maybe we also need a status like `Cancelled` or `Finished`.
Previously it was empty Block